design/generic_resources.md

```diff
@@ -0,0 +1,155 @@
+  * [Proposed Changes](#proposed-changes)
+
+##  Abstract
+
+This document describes the a solution to managing accountable node level
+resources unknown to docker swarm.
+
+## Motivation
+
+Each node is different in its own way, some nodes might have access to
+accelerators, some nodes might have access to network devices and others
+might support AVX while others only support SSE.
+Swarmkit needs some simple way to account for these resources without having
+to implement them each time a new kind of resource comes into existence.
+
+While it is true that some resources can be advertised with labels, many
+resources have a shareable capacity and can’t be represented well as a label.
+
+The implementation we chose is to reuse a proven solution used by industry
+projects (mesos and kubernetes) which lead us to implement two kinds
+of Generic resources:
+  * Discrete (uint64)
+  * Set
+  * Other types of resource like scalar can be extended
+
+Integer resources are for use cases where only an unsigned is needed to account
+for the resource (see Linux Realtime).
+
+A set would mostly be used for every resource which would need an
+exclusive access to it.
+
+## Constraints and Assumptions
+1. Future work might require new mechanisms to be made to allow Generic resources
+to be cluster wide in order to satisfy other use cases (e.g: pool of licenses)
+
+## Use Cases
+
+  * Exclusive access to discrete accelerators:
+    * GPU devices
+    * FPGA devices
+    * MICs (Many-Integrated Core, such as Xeon Phi)
+    * ...
+  * Support for tracking additional cgroup quotas like cpu_rt_runtime.
+    * [Linux Realtime](https://github.com/docker/docker/pull/23430)
+  * PersistentDisks in GCE
+  * Counting “slots” allowed access to a shared parallel file system.
+
+## Related Issues
+
+  * [Support abstract resource](https://github.com/docker/swarmkit/issues/594)
+  * [Add new node filter to scheduler](https://github.com/docker/swarm/issues/2223)
+  * [Add support for devices](https://github.com/docker/swarmkit/issues/1244)
+  * [Resource Control](https://github.com/docker/swarmkit/issues/211)
+  * [NVIDIA GPU support](https://github.com/docker/docker/issues/23917])
+  * [Does Docker have plan to support allocating GPU] (https://github.com/docker/docker/issues/24582)
+  * [Docker Swarm to orchestrate "Swarm Cluster" which supports GPU](https://github.com/docker/docker/issues/24750)
+  * [Use accelerator in docker container](https://github.com/docker/docker/issues/28642)
+  * [Specify resource selectors](https://github.com/docker/swarmkit/issues/206)
+
+## Objectives
+
+1. Associate multiple Generic resources with a node
+2. Request some portion of available Generic resourcess in the service
+   during service creation
+3. Enable users to define and schedule Generic resourcess in a vanilla swarmkit cluster
+
+## Non-Objectives
+
+1. Solve how Generic resourcess allocations are to be enforced or isolated.
+2. Solve how Generic resourcess are discovered
```

Comment:
typo: `resourcess` should be `resources`.

There are a number of places in this document where `Generic` is capitalized, but it probably shouldn't be.